pkg/controller/nstemplateset/nstemplateset_controller.go

```diff
@@ -0,0 +1,339 @@
+		}
+
+		// set labels
+		labels := acc.GetLabels()
+		if labels == nil {
+			labels = make(map[string]string)
+		}
+		labels["owner"] = username
+		labels["type"] = tcNamespace.Type
+		acc.SetLabels(labels)
+
+		// set owner ref
+		if err := controllerutil.SetControllerReference(nsTmplSet, acc, r.scheme); err != nil {
+			return r.wrapErrorWithStatusUpdate(logger, nsTmplSet, r.setStatusNamespaceProvisionFailed, err, "failed to set controller reference for namespace type '%s'", tcNamespace.Type)
+		}
+	}
+
+	err = tmplProcessor.Apply(objs)
+	if err != nil {
+		return r.wrapErrorWithStatusUpdate(logger, nsTmplSet, r.setStatusNamespaceProvisionFailed, err, "failed to create namespace with type '%s'", tcNamespace.Type)
+	}
+
+	log.Info("namespace provisioned", "namespace", tcNamespace)
+	return nil
+}
+
+func (r *ReconcileNSTemplateSet) ensureInnerNamespaceResources(logger logr.Logger, nsTmplSet *toolchainv1alpha1.NSTemplateSet, tcNamespace *toolchainv1alpha1.NSTemplateSetNamespace, params map[string]string, namespace *corev1.Namespace) error {
+	nsName := namespace.GetName()
+
+	tmplContent, err := r.getTemplateContent(nsTmplSet.Spec.TierName, tcNamespace.Type)
+	if err != nil {
+		return r.wrapErrorWithStatusUpdate(logger, nsTmplSet, r.setStatusNamespaceProvisionFailed, err, "failed to to retrieve template for namespace '%s'", nsName)
+	}
+
+	tmplProcessor := template.NewProcessor(r.client, r.scheme)
+	objs, err := tmplProcessor.Process(tmplContent, params, template.RetainAllButNamespaces)
+	if err != nil {
+		return r.wrapErrorWithStatusUpdate(logger, nsTmplSet, r.setStatusNamespaceProvisionFailed, err, "failed to process template for namespace '%s'", nsName)
+	}
+	err = tmplProcessor.Apply(objs)
+	if err != nil {
+		return r.wrapErrorWithStatusUpdate(logger, nsTmplSet, r.setStatusNamespaceProvisionFailed, err, "failed to provision namespace '%s'", nsName)
+	}
+
+	if namespace.Labels == nil {
+		namespace.Labels = make(map[string]string)
+	}
+	namespace.Labels["revision"] = tcNamespace.Revision
+	if err := r.client.Update(context.TODO(), namespace); err != nil {
+		return r.wrapErrorWithStatusUpdate(logger, nsTmplSet, r.setStatusNamespaceProvisionFailed, err, "failed to update namespace '%s'", nsName)
+	}
+
+	log.Info("namespace provisioned", "namespace", tcNamespace)
+
+	// TODO add validation for other objects
+	return nil
+}
+
+// nextNamespaceToProvision returns first namespace (from given namespaces) with
+// namespace status is active and revision not set
+// or namesapce present in tcNamespaces but not found in given namespaces
+func nextNamespaceToProvision(tcNamespaces []toolchainv1alpha1.NSTemplateSetNamespace, namespaces []corev1.Namespace) (*toolchainv1alpha1.NSTemplateSetNamespace, *corev1.Namespace, bool) {
+	for _, tcNamespace := range tcNamespaces {
+		namespace, found := findNamespace(namespaces, tcNamespace.Type)
+		if found {
+			if namespace.Status.Phase == corev1.NamespaceActive && namespace.GetLabels()["revision"] == "" {
```

Comment:
I think, you wanted to just verify that namespace has label revision here, How about using
 `!labels.Set(namespace.Labels).Has("revision")` from 	"k8s.io/apimachinery/pkg/labels" pkg.

Reply:
Done https://github.com/codeready-toolchain/member-operator/pull/53/commits/316e1ef2c52bb9f0d027054fd0e45e5a8942dcc0 rather I see `namespace.Labels["revision"] == ""` easy to read and understand.